### PR TITLE
JENKINS-45253 Creating a pipeline results in a not found error

### PR DESCRIPTION
**Pull request by i386**
I think this may solve a mystery where users can pick repositories but fail to save the Jenkinsfile. I was able to reproduce the problem as described in the ticket when I had myself changed to read only for a repository.

@vivek probably a stack of integration tests failing because I changed things.

# Description

See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-45253).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java [hunk 1]
```diff
@@ -66,6 +69,14 @@ public Object save(@Nonnull String apiUrl, @Nullable String owner, @Nullable Str
         }
 
         try {
+            // First validate that you have access to push to the repository
+            // Otherwise the call to post content will fail with a 404
+            GitHub gitHub = GitHub.connectToEnterprise(apiUrl, accessToken);
+            GHRepository repository = gitHub.getRepository(String.format("%s/%s", owner, repoName));
+            if (!(repository.hasAdminAccess() || repository.hasPushAccess())) {
```
Comment (kzantow) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java]: Shouldn't `hasPushAccess` be true if `hasAdminAccess`?

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubRepositories.java [hunk 2]
```diff
@@ -61,7 +62,8 @@ public GithubRepositories(StandardUsernamePasswordCredentials credentials, Strin
                     parent.getRepoType(),
                     pageSize, pageNumber), accessToken);
 
-            this.repositories = GithubScm.om.readValue(connection.getInputStream(), GHRepoEx[].class);
+            CollectionType collectionType = GithubScm.om.getTypeFactory().constructCollectionType(List.class, GHRepoEx.class);
```
Comment (kzantow) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubRepositories.java]: Could this be a constant?

Comment (i386) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubRepositories.java]: Done

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java (hunk 1 above)
Comment (i386) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java]: Yes, done.

**Comment by i386** [comment]
Thanks @kzantow

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java [hunk 3]
```diff
@@ -66,6 +69,14 @@ public Object save(@Nonnull String apiUrl, @Nullable String owner, @Nullable Str
         }
 
         try {
+            // First validate that you have access to push to the repository
+            // Otherwise the call to post content will fail with a 404
+            GitHub gitHub = GitHub.connectToEnterprise(apiUrl, accessToken);
```
Comment (vivek) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java]: We should do this check in [getContent()](https://github.com/jenkinsci-cert/blueocean-plugin/blob/c03b804b6ad81158c5908885750a7ea57b3621ca/blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmContentProvider.java#L40) as well. Better to fail fast than let user do all work then fail.

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java [hunk 4]
```diff
@@ -66,6 +69,14 @@ public Object save(@Nonnull String apiUrl, @Nullable String owner, @Nullable Str
...
+            GHRepository repository = gitHub.getRepository(String.format("%s/%s", owner, repoName));
+            if (!repository.hasPushAccess()) {
+                throw new ServiceException.ForbiddenException(String.format("You do not have permission to push changes to %s/%s", owner, repoName));
```
Comment (vivek) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java]: we should return 428 (PreconditionRequired) to be consistent.

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java (hunk 3 above)
Comment (i386) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java]: Done.

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java (hunk 4 above)
Comment (i386) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubScmSaveFileRequest.java]: Done.

**Pull request merged**